Orange/widgets/unsupervised/owtsne.py

```diff
@@ -1,60 +1,247 @@
 
-        # Larger data sets need a larger number of iterations
-        if self.n_samples > 100_000:
-            self.early_exagg_iter, self.n_iter = 500, 1000
-        else:
-            self.early_exagg_iter, self.n_iter = 250, 750
+            if state.is_interruption_requested():
+                raise InterruptRequested()
 
-    @property
-    def n_samples(self):
-        return self.embedding.embedding_.shape[0]
+    @classmethod
+    def run(cls, task, state):
+        # type: (Task, TaskState) -> Task
 
-    def run_optimization(self):
-        total_iterations = self.early_exagg_iter + self.n_iter
+        # Assign weights to each job indicating how much time will be spent on each
+        weights = {"pca": 1, "init": 1, "aff": 23, "tsne": 75}
 
-        # Default values of early exaggeration phase
-        exaggeration, momentum = 12, 0.5
+        # Prepare the tsne object and add it to the spec
+        task.tsne = prepare_tsne_obj(
+            task.data, task.perplexity, task.multiscale, task.exaggeration
+        )
 
-        current_iter = self.iterations_done
-        while not current_iter >= total_iterations:
-            # Switch to normal regime if early exaggeration phase is over
-            if current_iter >= self.early_exagg_iter:
-                exaggeration, momentum = self.exaggeration, 0.8
+        # Only execute jobs if needed
+        job_queue = []
+        if task.pca_projection is None:
+            job_queue.append(
+                (partial(cls.compute_pca, task, state), weights["pca"])
+            )
+        if task.initialization is None:
+            job_queue.append(
+                (partial(cls.compute_initialization, task, state), weights["init"])
+            )
+        if task.affinities is None:
+            job_queue.append(
+                (partial(cls.compute_affinities, task, state), weights["aff"])
+            )
 
-            # Resume optimization for some number of steps
-            self.embedding.optimize(
-                self.step_size, inplace=True, exaggeration=exaggeration,
-                momentum=momentum,
+        total_iterations = task.tsne.early_exaggeration_iter + task.tsne.n_iter
+        if task.tsne_embedding is None or task.iterations_done < total_iterations:
+            job_queue.append(
+                (partial(cls.compute_tsne, task, state), weights["tsne"])
             )
 
-            current_iter += self.step_size
+        # Figure out the total weight of the jobs
+        total_weight = sum(j[1] for j in job_queue)
+
+        try:
+            progress_done = 0
+            for job, job_weight in job_queue:
+                job_weight /= total_weight
+
+                def _progress_callback(val):
+                    state.set_progress_value(
+                        100 * progress_done + 100 * val * job_weight
+                    )
+
+                if state.is_interruption_requested():
+                    raise InterruptRequested()
```

Comment:
Why raising an exception here and excepting it a few lines below? Maybe you could just return the `task`.

Reply:
Removed the exception altogether.